File: LD31Dart/web/ParticleEmitter.dart
library ld31_particleEmitter;

import "dart:html";
import "dart:math";
import "Particle.dart";

class ParticleEmitter{
  
  List<Particle> _particles;
  List<int> _toRemove;
  
  double _x, _y;
  String _style;
  int _amount, _generated, _removed;
  int _particleWidth, _particleHeight;
  int _minSpeed, _maxSpeed, _minAngle, _maxAngle, _minLife, _maxLife;
  
  ParticleEmitter(double x, double y, int amount, int particleWidth, int particleHeight, String style,
      int minSpeed, int maxSpeed, int minAngle, int maxAngle, int minLife, int maxLife){
    _particles = [];
    _toRemove = [];
    
    _x = x;
    _y = y;
    _amount = amount;
    _particleWidth = particleWidth;
    _particleHeight = particleHeight;
    _style = style;
    _minSpeed = minSpeed;
    _maxSpeed = maxSpeed;
    _minAngle = minAngle;
    _maxAngle = maxAngle;
    _minLife = minLife;
    _maxLife = maxLife;
    
    _generated = 0;
    _removed = 0;
  }
  
  void tick(final double delta){
    if(_generated < _amount){
      _emit();
      _generated++;
    }
    
    for(int i = 0;i < _particles.length;++i){
      if(_particles[i].tick(delta))
        _toRemove.add(i);
    }
    
    for(int i = 0;i < _toRemove.length;++i){
      if(_toRemove[i] < _particles.length)
        _particles.removeAt(_toRemove[i]);
    }
    _toRemove.clear();
  }
  
  void render(CanvasRenderingContext2D g){
    g.fillStyle = _style;
    
    for(int i = 0;i < _particles.length;++i)
      _particles[i].render(g, _particleWidth, _particleHeight);
  }
  
  void _emit(){
    _particles.add(new Particle(_x, _y, _randomize(_minAngle, _maxAngle), _randomize(_minSpeed, _maxSpeed), _randomize(_minLife, _maxLife)));
  }
  
  int _randomize(int min, int max){
    var r = new Random();
    return r.nextInt((max - min) + 1) + min;
  }
  
}

File: LD31/src/competition/codenmore/ld31/particles/ParticleEmitter.java
package competition.codenmore.ld31.particles;

import java.awt.Color;
import java.awt.Graphics;
import java.util.ArrayList;
import java.util.Random;

public class ParticleEmitter {
	
	private Random random;
	
	private float x, y;
	private int width = 4, height = 4;
	private int minSpeed, maxSpeed, minAngle, maxAngle, minLife, maxLife;
	private int amount, generated = 0, removed = 0;
	private Color color;
	private long last, now;
	private boolean swap = false;
	
	private ArrayList<Particle> particles;

	public ParticleEmitter(float x, float y, int amount, int minSpeed, int maxSpeed, int minAngle, int maxAngle, int minLife, int maxLife, Color color){
		this.x = x;
		this.y = y;
		
		this.amount = amount;
		this.minSpeed = minSpeed;
		this.maxSpeed = maxSpeed;
		this.minAngle = minAngle;
		this.maxAngle = maxAngle;
		this.minLife = minLife;
		this.maxLife = maxLife;
		this.color = color;
		particles = new ArrayList<Particle>();
		
		random = new Random();
		last = System.currentTimeMillis();
	}
	
	public boolean tick(){
		swap = !swap;
		
		if(swap && generated <= amount){
			emit();
			generated++;
		}
		
		now = System.currentTimeMillis();
		
		for(Particle p : particles){
			if(p.isActive() && p.tick((int) (now - last))){
				removed++;
				p.setActive(false);
			}
		}
		
		last = now;
		
		if(removed >= amount)
			return true;
		
		return false;
	}
	
	public void render(Graphics g){
		g.setColor(color);
		
		for(Particle p : particles)
			if(p.isActive())
				p.render(g);
	}
	
	public void emit(){
		particles.add(new Particle(x, y, randomize(minSpeed, maxSpeed), randomize(minAngle, maxAngle), randomize(minLife, maxLife), this));
	}
	
	private int randomize(int min, int max){
		return random.nextInt((max - min) + 1) + min;
	}

	public float getX() {
		return x;
	}

	public void setX(float x) {
		this.x = x;
	}

	public float getY() {
		return y;
	}

	public void setY(float y) {
		this.y = y;
	}

	public int getWidth() {
		return width;
	}

	public void setWidth(int width) {
		this.width = width;
	}

	public int getHeight() {
		return height;
	}

	public void setHeight(int height) {
		this.height = height;
	}

	public int getMinSpeed() {
		return minSpeed;
	}

	public void setMinSpeed(int minSpeed) {
		this.minSpeed = minSpeed;
	}

	public int getMaxSpeed() {
		return maxSpeed;
	}

	public void setMaxSpeed(int maxSpeed) {
		this.maxSpeed = maxSpeed;
	}

	public int getAmount() {
		return amount;
	}

	public void setAmount(int amount) {
		this.amount = amount;
	}

	public Color getColor() {
		return color;
	}

	public void setColor(Color color) {
		this.color = color;
	}
	
}
